synapse/rest/client/transactions.py

```diff
@@ -79,18 +77,11 @@ def _get_transaction_key(self, request: IRequest, requester: Requester) -> Hasha
             return (path, "appservice", requester.app_service.id)
 
         # With MSC3970, we use the user ID and device ID as the transaction key
-        elif self._msc3970_enabled:
+        else:
             assert requester.user, "Requester must have a user"
             assert requester.device_id, "Requester must have a device_id"
             return (path, "user", requester.user, requester.device_id)
 
-        # Otherwise, the pre-MSC3970 behaviour is to use the access token ID
-        else:
-            assert (
-                requester.access_token_id is not None
-            ), "Requester must have an access_token_id"
-            return (path, "user", requester.access_token_id)
```

Comment:
Or should we be checking if the device ID exists above?

Reply:
The only users that don't have a device ID are:

 - appservices
 - guests
 - access tokens generated by the admin API

Only the last case might go in this branch, and since it's un-specced, rare behavior, and it doesn't make the request explode (since asserts are I think disabled in production?), I wouldn't bother 🤷